kvbc/src/kvbc_app_filter/kvbc_app_filter.cpp

```diff
@@ -355,236 +354,197 @@ uint64_t KvbAppFilter::newestTagSpecificPublicEventGroupId() const {
-  if (private_end == 0 && eg_state->public_offset > public_end) return std::nullopt;
-  // no public and private event groups in storage
-  if (public_end == 0 && private_end == 0) return std::nullopt;
-  // all public and private event group ids have already been read
-  if (eg_state->private_offset > private_end && eg_state->public_offset > public_end) return std::nullopt;
-
-  // reset the vectors to read the new batch from storage
-  eg_state->event_group_id_batch.clear();
-  eg_state->it = eg_state->event_group_id_batch.end();
-  // holds a maximum of kBatchSize ordered public event group ids at a time
-  std::vector<uint64_t> public_event_group_ids{};
-  // holds a maximum of kBatchSize ordered private event group ids at a time
-  std::vector<uint64_t> private_event_group_ids{};
-
-  // populate public_event_group_ids
-  if (eg_state->public_offset != 0) {
-    // start reading from the offset until a complete batch is read or no more event groups exist for the tag
-    for (uint64_t i = eg_state->public_offset; i < std::min(eg_state->public_offset + kBatchSize, public_end + 1);
-         ++i) {
-      // get global_event_group_id and external_tag_event_group_id corresponding to tag_event_group_id
-      // tag + kTagTableKeySeparator + latest_tag_event_group_id concatenation is used as key for kv-updates of type
-      // kExecutionEventGroupTagCategory
-      const auto &[global_eg_id, _] =
-          getValueFromTagTable(kPublicEgId + kTagTableKeySeparator + concordUtils::toBigEndianStringBuffer(i));
-      (void)_;
-      public_event_group_ids.emplace_back(global_eg_id);
-    }
-  }
-  // populate private_event_group_ids
-  if (eg_state->private_offset != 0) {
-    for (uint64_t i = eg_state->private_offset; i < std::min(eg_state->private_offset + kBatchSize, private_end + 1);
-         ++i) {
-      const auto &[global_eg_id, _] =
-          getValueFromTagTable(client_id_ + kTagTableKeySeparator + concordUtils::toBigEndianStringBuffer(i));
-      (void)_;
-      private_event_group_ids.emplace_back(global_eg_id);
+
+  if (not private_start) {
+    // requested external event group id == public event group id
+    uint64_t global_id;
+    std::tie(global_id, std::ignore) = getValueFromTagTable(
+        kPublicEgId + kTagTableKeySeparator + concordUtils::toBigEndianStringBuffer(external_event_group_id));
+    return {global_id, false, private_end ? private_end - 1 : 0, external_event_group_id};
+  } else if (not public_start) {
+    // requested external event group id == private event group id
+    uint64_t global_id;
+    std::tie(global_id, std::ignore) = getValueFromTagTable(
+        client_id_ + kTagTableKeySeparator + concordUtils::toBigEndianStringBuffer(external_event_group_id));
+    return {global_id, true, external_event_group_id, public_end ? public_end - 1 : 0};
+  }
+
+  std::string prefix = client_id_ + kTagTableKeySeparator;
+
+  // Binary search in private event groups
+  uint64_t window_size;
```

Comment:
Nit: Does it make sense to compute the `window_size` here and then update it at the end of the `while` loop below (instead of the beginning)? Doing it that way, we will ensure it is always initialized.

Reply:
The idea is to only have one place in which we compute the `window_size` which makes reasoning a tiny bit easier IMHO. The variable is only used inside the while loop and the first statement sets the value. Let me know if it this doesn't make sense to you and we can change it.

Reply:
The idea is to only have one place in which we compute the `window_size` which makes reasoning a tiny bit easier IMHO. The variable is only used inside the while loop and the first statement sets the value. Let me know if it this doesn't make sense to you and we can change it.